ArborX fails to compile with CUDA 11.0.3-11.3.3

Compiling `benchmarks/dbscan` results in 
```
nvcc error   : 'cicc' died due to signal 11 (Invalid memory reference)
nvcc error   : 'cicc' core dumped
make[2]: *** [benchmarks/dbscan/CMakeFiles/ArborX_Benchmark_DBSCAN.exe.dir/build.make:76: benchmarks/dbscan/CMakeFiles/ArborX_Benchmark_DBSCAN.exe.dir/dbscan_3.cpp.o] Error 139
```

Compiling `examples/dbscan` results in
```
<snip>/arborx/src/ArborX_DBSCAN.hpp: In instantiation of 'Kokkos::View<int*, typename ArborX::AccessTraits<Primitives, ArborX::PrimitivesTag, void>::memory_space> ArborX::dbscan(const ExecutionSpace&, const Primitives&, float, int, const ArborX::DBSCAN::Parameters&) [with ExecutionSpace = Kokkos::Cuda; Primitives = Kokkos::View<ArborX::Point*, Kokkos::CudaSpace>; typename ArborX::AccessTraits<Primitives, ArborX::PrimitivesTag>::memory_space = Kokkos::CudaSpace]':
<snip>/arborx/examples/dbscan/example_dbscan.cpp:59:64:   required from here
<snip>/arborx/src/ArborX_DBSCAN.hpp:301:14: error: 'HalfTraversal' was not declared in this scope; did you mean 'ArborX::Details::HalfTraversal'?
  301 |       Details::HalfTraversal(
      |       ~~~~~~~^~~~~~~~~~~~~~~~
      | ArborX::Details::HalfTraversal
<snip>/arborx/src/details/ArborX_DetailsHalfTraversal.hpp:24:8: note: 'ArborX::Details::HalfTraversal' declared here
   24 | struct HalfTraversal
      |        ^~~~~~~~~~~~~
<snip>/arborx/src/ArborX_DBSCAN.hpp:323:14: error: 'HalfTraversal' was not declared in this scope; did you mean 'ArborX::Details::HalfTraversal'?
  323 |       Details::HalfTraversal(exec_space, bvh,
      |       ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                   
      | ArborX::Details::HalfTraversal
<snip>/arborx/src/details/ArborX_DetailsHalfTraversal.hpp:24:8: note: 'ArborX::Details::HalfTraversal' declared here
   24 | struct HalfTraversal
      |        ^~~~~~~~~~~~~
nvcc_internal_extended_lambda_implementation:991:13: error: 'static decltype (typename __nv_hdl_helper_trait_outer<IsMutable, HasFuncPtrConv, CaptureArgs ...>::__nv_hdl_helper_trait<Tag, Lambda>::get(lam, __nv_hdl_create_wrapper_t::__nv_hdl_create_wrapper::args ...)) __nv_hdl_create_wrapper_t<IsMutable, HasFuncPtrConv, Tag, CaptureArgs>::__nv_hdl_create_wrapper(Lambda&&, CaptureArgs ...) [with Lambda = ArborX::dbscan(const ExecutionSpace&, const Primitives&, float, int, const ArborX::DBSCAN::Parameters&) [with ExecutionSpace = Kokkos::Cuda; Primitives = Kokkos::View<ArborX::Point*, Kokkos::CudaSpace>; typename ArborX::AccessTraits<Primitives, ArborX::PrimitivesTag>::memory_space = Kokkos::CudaSpace]::<lambda(int)>; bool IsMutable = false; bool HasFuncPtrConv = false; Tag = __nv_dl_tag<Kokkos::View<int*, Kokkos::CudaSpace> (*)(const Kokkos::Cuda&, const Kokkos::View<ArborX::Point*, Kokkos::CudaSpace>&, float, int, const ArborX::DBSCAN::Parameters&), ArborX::dbscan<Kokkos::Cuda, Kokkos::View<ArborX::Point*, Kokkos::CudaSpace> >, 1>; CaptureArgs = {Kokkos::View<int*, Kokkos::CudaSpace>, Kokkos::View<unsigned int*, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace> >}; decltype (typename __nv_hdl_helper_trait_outer<IsMutable, HasFuncPtrConv, CaptureArgs ...>::__nv_hdl_helper_trait<Tag, Lambda>::get(lam, __nv_hdl_create_wrapper_t::__nv_hdl_create_wrapper::args ...)) = __nv_hdl_wrapper_t<false, false, __nv_dl_tag<Kokkos::View<int*, Kokkos::CudaSpace> (*)(const Kokkos::Cuda&, const Kokkos::View<ArborX::Point*, Kokkos::CudaSpace>&, float, int, const ArborX::DBSCAN::Parameters&), ArborX::dbscan<Kokkos::Cuda, Kokkos::View<ArborX::Point*, Kokkos::CudaSpace> >, 1>, void(int), Kokkos::View<int*, Kokkos::CudaSpace>, Kokkos::View<unsigned int*, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace> > >; typename __nv_hdl_helper_trait_outer<IsMutable, HasFuncPtrConv, CaptureArgs ...>::__nv_hdl_helper_trait<Tag, Lambda> = __nv_hdl_helper_trait_outer<false, false, Kokkos::View<int*, Kokkos::CudaSpace>, Kokkos::View<unsigned int*, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace> > >::__nv_hdl_helper_trait<__nv_dl_tag<Kokkos::View<int*, Kokkos::CudaSpace> (*)(const Kokkos::Cuda&, const Kokkos::View<ArborX::Point*, Kokkos::CudaSpace>&, float, int, const ArborX::DBSCAN::Parameters&), ArborX::dbscan<Kokkos::Cuda, Kokkos::View<ArborX::Point*, Kokkos::CudaSpace> >, 1>, ArborX::dbscan(const ExecutionSpace&, const Primitives&, float, int, const ArborX::DBSCAN::Parameters&) [with ExecutionSpace = Kokkos::Cuda; Primitives = Kokkos::View<ArborX::Point*, Kokkos::CudaSpace>; typename ArborX::AccessTraits<Primitives, ArborX::PrimitivesTag>::memory_space = Kokkos::CudaSpace]::<lambda(int)> >]', declared using local type 'ArborX::dbscan(const ExecutionSpace&, const Primitives&, float, int, const ArborX::DBSCAN::Parameters&) [with ExecutionSpace = Kokkos::Cuda; Primitives = Kokkos::View<ArborX::Point*, Kokkos::CudaSpace>; typename ArborX::AccessTraits<Primitives, ArborX::PrimitivesTag>::memory_space = Kokkos::CudaSpace]::<lambda(int)>', is used but never defined [-fpermissive]
```

--- a/benchmarks/dbscan/data.hpp
+++ b/benchmarks/dbscan/data.hpp
@@ -27,7 +27,9 @@ std::vector<Point<DIM>> sampleData(std::vector<Point<DIM>> const &data,
   std::vector<Point<DIM>> sampled_data(num_samples);
 
   std::default_random_engine generator(17);
-  std::uniform_int_distribution distribution;
+  // FIXME_NVCC(11.0-11.3): not providing int here results in an internal
+  // compiler error
+  std::uniform_int_distribution<int> distribution;
   auto rand = [&distribution, &generator]() { return distribution(generator); };
 
   // Knuth algorithm

--- a/src/ArborX_DBSCAN.hpp
+++ b/src/ArborX_DBSCAN.hpp
@@ -297,10 +297,14 @@ dbscan(ExecutionSpace const &exec_space, Primitives const &primitives,
     {
       // Perform the queries and build clusters through callback
       using CorePoints = Details::CCSCorePoints;
-      // FIXME_NVCC(11.0-11.3): this HalfTraversal typedef is a workaround
+#if defined(KOKKOS_COMPILER_NVCC) && (KOKKOS_COMPILER_NVCC < 1140)
+      // Workaround a compiler bug
       using HalfTraversal = Details::HalfTraversal<
           decltype(bvh), Details::FDBSCANCallback<UnionFind, CorePoints>,
           Details::WithinRadiusGetter>;
+#else
+      using Details::HalfTraversal;
+#endif
       Kokkos::Profiling::pushRegion("ArborX::DBSCAN::clusters::query");
       HalfTraversal(
           exec_space, bvh,
@@ -321,10 +325,14 @@ dbscan(ExecutionSpace const &exec_space, Primitives const &primitives,
       Kokkos::Profiling::popRegion();
 
       using CorePoints = Details::DBSCANCorePoints<MemorySpace>;
-      // FIXME_NVCC(11.0-11.3): this HalfTraversal typedef is a workaround
+#if defined(KOKKOS_COMPILER_NVCC) && (KOKKOS_COMPILER_NVCC < 1140)
+      // Workaround a compiler bug
       using HalfTraversal = Details::HalfTraversal<
           decltype(bvh), Details::FDBSCANCallback<UnionFind, CorePoints>,
           Details::WithinRadiusGetter>;
+#else
+      using Details::HalfTraversal;
+#endif
 
       // Perform the queries and build clusters through callback
       Kokkos::Profiling::pushRegion("ArborX::DBSCAN::clusters::query");

--- a/src/ArborX_DBSCAN.hpp
+++ b/src/ArborX_DBSCAN.hpp
@@ -297,8 +297,12 @@ dbscan(ExecutionSpace const &exec_space, Primitives const &primitives,
     {
       // Perform the queries and build clusters through callback
       using CorePoints = Details::CCSCorePoints;
+      // FIXME_NVCC(11.0-11.3): this HalfTraversal typedef is a workaround
+      using HalfTraversal = Details::HalfTraversal<
+          decltype(bvh), Details::FDBSCANCallback<UnionFind, CorePoints>,
+          Details::WithinRadiusGetter>;
       Kokkos::Profiling::pushRegion("ArborX::DBSCAN::clusters::query");
-      Details::HalfTraversal(
+      HalfTraversal(
           exec_space, bvh,
           Details::FDBSCANCallback<UnionFind, CorePoints>{labels, CorePoints{}},
           Details::WithinRadiusGetter{eps});
@@ -317,13 +321,17 @@ dbscan(ExecutionSpace const &exec_space, Primitives const &primitives,
       Kokkos::Profiling::popRegion();
 
       using CorePoints = Details::DBSCANCorePoints<MemorySpace>;
+      // FIXME_NVCC(11.0-11.3): this HalfTraversal typedef is a workaround
+      using HalfTraversal = Details::HalfTraversal<
+          decltype(bvh), Details::FDBSCANCallback<UnionFind, CorePoints>,
+          Details::WithinRadiusGetter>;
 
       // Perform the queries and build clusters through callback
       Kokkos::Profiling::pushRegion("ArborX::DBSCAN::clusters::query");
-      Details::HalfTraversal(exec_space, bvh,
-                             Details::FDBSCANCallback<UnionFind, CorePoints>{
-                                 labels, CorePoints{num_neigh, core_min_size}},
-                             Details::WithinRadiusGetter{eps});
+      HalfTraversal(exec_space, bvh,
+                    Details::FDBSCANCallback<UnionFind, CorePoints>{
+                        labels, CorePoints{num_neigh, core_min_size}},
+                    Details::WithinRadiusGetter{eps});
       Kokkos::Profiling::popRegion();
     }
   }

